### ISSUE TITLE
"cookie not present" on iOS but works with all other platforms

### ISSUE THREAD

**Issue by sneko** [oauth2-proxy/oauth2-proxy#830]
## Issue

The proxy successfully works for my desktops devices and Android ones but not for iOS.

I always get this error in the frontend:
```
403 Permission Denied

http: named cookie not present
```

I also took a look at proxy logs and here they are:
```
10.24.0.22:48810 - XXXXXXX@XXXXXXXXX [2020/10/08 14:23:18] [AuthFailure] Invalid authentication via OAuth2: unable to obtain CSRF cookie
...
[2020/10/08 14:23:19] [logger.go:508] Error loading cookied session: cookie "_oauth2_proxy" not present, removing session
```

I had issue 2 years ago with my iPhone blocking cross-domain cookies and so on, but everything was solved and I rechecked my system settings. (Cookies are not blocked also)

I tried in:
- iOS Safari
- iOS Brave
- iOS Chrome

But also on my phone network, and on the WiFi (just in case...)

I also had the following settings for the proxy to fit new security browsers requirements (put them a few months ago): 
```
        - --cookie-httponly=true
        - --cookie-secure=true
        - --cookie-samesite=none
```

And just to be sure I removed them in case iPhone had weird behavior with them but it's also not working.

I don't think it comes from the proxy, but I hope someone can help 👍 

Thank you,

## Your Environment

- Version used: v6.1.1 (but also tested on v6.0.0)

**Comment by sneko** [comment]
Just found what's making things blocking! Pfffiouuuuuu 😆 

Instead of using:
```
        - --cookie-httponly=true
        - --cookie-secure=true
        - --cookie-samesite=none
```

I have to use:
```
        - --cookie-httponly=true
        - --cookie-secure=true
        - --cookie-samesite=lax
```

Initially I set `samesite=none` because I do some cross-subdomains requests. From what I understood it was the only way to reach my other subdomains even if the cookie was on parent domain (.domain.com) because `samesite` setting is about the URL domain, not the cookie domain itself.

But it seems I was wrong, or that I missed something.

Anyway, I hope it will help others, because I just found other people talking about this mess... and I guess Apple will still do nothing to fix previous iOS (even if my iPhone 6 is not that old):
- https://developer.apple.com/forums/thread/129064
- https://stackoverflow.com/questions/58525719/safari-not-sending-cookie-even-after-setting-samesite-none-secure
- https://www.chromium.org/updates/same-site/incompatible-clients

I hope using "lax" instead of "none" won't break any of my tools... otherwise in above articles they advise to set the `samesite` value depending on the user agent, sounds like a big overhead 👎 

@JoelSpeed I don't know if a mention is needed in the repository, or at least my issue can help knowing about this.

(closing since not on your hand)

Thank you for your work,

**Comment by sneko** [comment]
Haha @JoelSpeed  short-lived happiness... It works for production and so on but not when developing locally with my `localhost:XXXX`.

I guess I have to deploy 2 proxies, one for all devices, and a specific for Safari (iOS/Mac). I just need to figure out how to route to the correct one depending on the user agent.

The other solution would be the proxy to also some sort of routing due to user agent... but I don't think you will agree with that (me neither by the way...).

**Comment by sneko** [comment]
Hi,

So I worked around this big Apple issue...

- For a basic browser on iOS I modified the `oauth2-proxy` so it overrides it to `lax` when `none` is used
- On my native application I had to set 2 years ago the cookie manually since the URL of the app is like `yourScheme://localhost/path` but it was no longer working even with the `oauth2-proxy` modification.

I wanted to add `SameSite=lax` natively but customizing it is for iOS 13 at least, and I'm stuck on iOS 12... It's hard for me to consider telling to all testers in the company "we have to upgrade our phones...". So the workaround I did:

Instead of putting a cookie on a parent domain like `.domain.com`, I made a loop on all subdomains needed `www | api | example | ...` (that's an example). Here how the code:

``` 
        let subdomainsToSetCookieOn = ["www", "api", "example"]
        for subdomain in subdomainsToSetCookieOn {
            let cookie = HTTPCookie(properties: [
                HTTPCookiePropertyKey.domain: subdomain+".domain.com",
                HTTPCookiePropertyKey.path: "/",
                HTTPCookiePropertyKey.name: "_oauth2_proxy",
                HTTPCookiePropertyKey.value: accessToken!,
                HTTPCookiePropertyKey.maximumAge: 60 * 60 * 24 * 30 * 6, // Keep the cookie 6 months
                HTTPCookiePropertyKey.secure: "Yes",
            ])!
            HTTPCookieStorage.shared.setCookie(cookie);
        }
```
_You need to put this code where you handle the original cookie value... for me it was by injecting the value at application start through a deeplink (universal link)_

I hope it will help someone :) ... That's still hacky, but after seeing 392494 posts of people desesperate about this Apple issue, I think it worth it 😄 

🥳
